app/assets/javascripts/oxalis/model/helpers/nml_helpers.js

```diff
@@ -0,0 +1,194 @@
+        xRot: editRotation[0],
+        yRot: editRotation[1],
+        zRot: editRotation[2],
+      }),
+      serializeTag("zoomLevel", { zoom: state.flycam.zoomStep }),
+      userBB != null
+        ? serializeTag("userBoundingBox", {
+            topLeftX: userBB.min[0],
+            topLeftY: userBB.min[1],
+            topLeftZ: userBB.min[2],
+            width: userBB.max[0] - userBB.min[0],
+            height: userBB.max[1] - userBB.min[1],
+            depth: userBB.max[2] - userBB.min[2],
+          })
+        : "",
+    ]),
+    "</parameters>",
+  ];
+}
+
+function serializeTrees(trees: Array<TreeType>): Array<string> {
+  return _.flatten(
+    trees.map(tree => [
+      serializeTag(
+        "thing",
+        {
+          id: tree.treeId,
+          "color.r": tree.color[0],
+          "color.g": tree.color[1],
+          "color.b": tree.color[2],
+          "color.a": 1.0,
+          name: tree.name,
+        },
+        false,
+      ),
+      ...indent([
+        "<nodes>",
+        ...indent(serializeNodes(tree.nodes)),
+        "</nodes>",
+        "<edges>",
+        ...indent(serializeEdges(tree.edges)),
+        "</edges>",
+      ]),
+      "</thing>",
+    ]),
+  );
+}
+
+function serializeNodes(nodes: NodeMapType): Array<string> {
+  return Object.keys(nodes).map(nodeId => {
+    const node = nodes[+nodeId];
+    return serializeTag("node", {
+      id: node.id,
+      radius: node.radius,
+      x: node.position[0],
+      y: node.position[1],
+      z: node.position[2],
+      rotX: node.rotation[0],
+      rotY: node.rotation[1],
+      rotZ: node.rotation[2],
+      inVp: node.viewport,
+      inMag: node.resolution,
+      bitDepth: node.bitDepth,
+      interpolation: node.interpolation,
+      time: node.timestamp,
+    });
+  });
+}
+
+function serializeEdges(edges: Array<EdgeType>): Array<string> {
+  return edges.map(edge => serializeTag("edge", { source: edge.source, target: edge.target }));
+}
+
+function serializeBranchPoints(trees: Array<TreeType>): Array<string> {
+  const branchPoints = _.flatten(trees.map(tree => tree.branchPoints));
+  return [
+    "<branchpoints>",
+    ...indent(
+      branchPoints.map(branchPoint =>
+        serializeTag("branchpoint", { id: branchPoint.nodeId, time: branchPoint.timestamp }),
+      ),
+    ),
+    "</branchpoints>",
+  ];
+}
+
+function serializeComments(trees: Array<TreeType>): Array<string> {
+  const comments = _.flatten(trees.map(tree => tree.comments));
+  return [
+    "<comments>",
+    ...indent(
+      comments.map(comment =>
+        serializeTag("comment", { node: comment.nodeId, content: comment.content }),
+      ),
+    ),
+    "</comments>",
+  ];
+}
+
+function serializeTag(
+  name: string,
+  properties: { [string]: ?(string | number | boolean) },
+  closed: boolean = true,
+): string {
+  return `<${name} ${Object.keys(properties)
+    .map(key => `${key}="${properties[key] != null ? properties[key].toString() : ""}"`)
```

Comment:
Object to string. Perhaps`JSON.stringify`? BTW, stringify accepts "indentation".

Reply:
Not sure I understand the benefit here. `JSON.stringify` would yield sth. like `{"name":"test","x":3}`, whereas I am building sth. like `<tag name="test" x="3">`.

Reply:
Right, that's probably not useful here.